File: src/ui/widgets/zoom_indicator.py
from __future__ import annotations

from typing import Callable

from PyQt6.QtCore import QSize, Qt
from PyQt6.QtGui import QColor
from PyQt6.QtWidgets import QHBoxLayout, QLabel, QWidget

from resources.translations import tr
from sli_ui_toolkit.widgets import Button
from ui.icon_manager import AppIcon
from ui.widgets.rounded_overlay import RoundedOverlayWidget


class ZoomIndicator(RoundedOverlayWidget):
    """Overlay showing current zoom percent with a reset button.

    Self-contained: takes a `lang_provider` callback to localize its prefix.
    The owning UI is responsible for connecting `btn_zoom_reset.clicked`.
    """

    def __init__(
        self,
        parent: QWidget,
        *,
        lang_provider: Callable[[], str] = lambda: "en",
        target_widget: QWidget | None = None,
    ):
        super().__init__(parent, bg_color=QColor(0, 0, 0, 140), radius=6)
        self._lang_provider = lang_provider
        self._target_widget = target_widget

        self.setObjectName("ZoomIndicator")
        layout = QHBoxLayout(self)
        layout.setContentsMargins(6, 2, 4, 2)
        layout.setSpacing(4)

        self._label = QLabel("100%", self)
        layout.addWidget(self._label)

        self.btn_zoom_reset = Button(AppIcon.SYNC, parent=self)
        self.btn_zoom_reset.setFixedSize(QSize(22, 22))
        self.btn_zoom_reset.setToolTip("Reset zoom")
        layout.addWidget(self.btn_zoom_reset)

        self.adjustSize()
        self.hide()
        self.update_zoom(1.0)

    def set_target(self, target: QWidget):
        self._target_widget = target

    def update_zoom(self, zoom: float, pan_x: float = 0.0, pan_y: float = 0.0):
        percent = int(round(float(zoom) * 100))
        prefix = tr("label.zoom", self._lang_provider())
        self._label.setText(f"{prefix}: {percent}%")
        self.adjustSize()
        visible = (
            abs(float(zoom) - 1.0) > 1e-3
            or abs(float(pan_x)) > 1e-4
            or abs(float(pan_y)) > 1e-4
        )
        self.setVisible(visible)
        if visible:
            self.sync_position()
            self.raise_()

    def sync_position(self):
        if not self.isVisible() or self._target_widget is None:
            return
        target_geo = self._target_widget.geometry()
        margin = 8
        w = self.width()
        h = self.height()
        x = target_geo.right() - w - margin
        y = target_geo.top() + margin
        self.move(x, y)
        self.raise_()

File: cpp/app/ui/widgets/zoom_indicator.cpp
#include "ui/widgets/zoom_indicator.h"

#include "ui/icon_manager.h"

#include <QHBoxLayout>
#include <QLabel>
#include <QPushButton>
#include <QSize>

#include <cmath>

namespace imgsli::app::ui::widgets {

ZoomIndicator::ZoomIndicator(QWidget* parent, PrefixProvider prefixProvider,
                             QWidget* target)
    : RoundedOverlayWidget(parent, QColor(0, 0, 0, 140), 6.0),
      prefixProvider_(std::move(prefixProvider)),
      target_(target) {
  setObjectName(QStringLiteral("ZoomIndicator"));
  auto* layout = new QHBoxLayout(this);
  layout->setContentsMargins(6, 2, 4, 2);
  layout->setSpacing(4);

  label_ = new QLabel(QStringLiteral("100%"), this);
  layout->addWidget(label_);

  resetButton_ = new QPushButton(this);
  resetButton_->setIcon(getAppIcon(AppIcon::Sync));
  resetButton_->setFixedSize(QSize(22, 22));
  resetButton_->setToolTip(QStringLiteral("Reset zoom"));
  resetButton_->setFlat(true);
  layout->addWidget(resetButton_);

  adjustSize();
  hide();
  updateZoom(1.0);
}

void ZoomIndicator::setTarget(QWidget* target) {
  target_ = target;
  syncPosition();
}

void ZoomIndicator::updateZoom(double zoom, double panX, double panY) {
  const int percent = static_cast<int>(std::round(zoom * 100.0));
  const QString prefix = prefixProvider_ ? prefixProvider_() : QStringLiteral("Zoom");
  label_->setText(QStringLiteral("%1: %2%").arg(prefix).arg(percent));
  adjustSize();

  const bool visible =
      std::abs(zoom - 1.0) > 1e-3 || std::abs(panX) > 1e-4 || std::abs(panY) > 1e-4;
  setVisible(visible);
  if (visible) {
    syncPosition();
    raise();
  }
}

void ZoomIndicator::syncPosition() {
  if (!isVisible() || !target_) {
    return;
  }
  const QRect targetGeo = target_->geometry();
  constexpr int margin = 8;
  const int x = targetGeo.right() - width() - margin;
  const int y = targetGeo.top() + margin;
  move(x, y);
  raise();
}

}  // namespace imgsli::app::ui::widgets
